src/WooCommerce.php

```diff
@@ -51,11 +51,11 @@ public static function woocommerce_add_to_cart($cart_item_key, $product_id, $qua
    *
    * @implements woocommerce_before_single_product
    */
-  public static function woocommerce_before_single_product() {
-    if (!static::$isAddedToCart || !$related_accessories_ids = static::getRelatedAccessoriesIds()) {
+  public static function woocommerce_before_single_product($force_render = FALSE) {
+    $related_accessories_ids = static::getRelatedAccessoriesIds();
```

Comment:
This fetches the related accessories even when they are not used. They should only be fetched when they will be used.

Reply:
@sun can you be more specific? When would they not be used? In my understanding, it is always used on the single product template.

Reply:
If they would always be rendered on the single product template, why have any output condition at all?

I think the additional `$force_render` parameter makes the output logic here unnecessarily complex.  I'd advise to split the rendering into an own function, so you can force-generate the output without having to deal with the regular page output logic: (see below)